### PR TITLE
Assert one-shot completion in IndexShardOperationPermits

### PR DESCRIPTION
**Pull request by DaveCTurner**
It's vital that listeners passed to the `IndexShardOperationPermits` are not completed more than once, because otherwise the caller might not release the returned `Releasable` which would leak a permit. It looks like this is the case today, but one of its callers has some explicit protection against such a leak. This commit adds some assertions to demonstrate that this protection is not required.

### CODE REVIEW

**Review by fcofdez — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/action/ActionListener.java [hunk 1]
```diff
@@ -532,20 +532,27 @@ public String toString() {
...
         if (Assertions.ENABLED) {
             return new ActionListener<>() {
-                private final AtomicBoolean alreadyRan = new AtomicBoolean();
+
+                // if complete, records the stack trace which first completed it
+                private final AtomicReference<ElasticsearchException> firstCompletion = new AtomicReference<>();
+
+                private void assertFirstRun() {
+                    var previousRun = firstCompletion.compareAndExchange(null, new ElasticsearchException(delegate.toString()));
+                    assert previousRun == null : previousRun; // reports the stack traces of both completions
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/action/ActionListener.java:544]: :+1:

**Review by tlrx — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
@elasticmachine update branch

**Pull request merged**